src/debug/debug_impl.ts

```diff
@@ -0,0 +1,197 @@
+'use strict';
+
+import * as child_process from "child_process";
+import * as path from "path";
+import {
+	DebugSession,
+	InitializedEvent, TerminatedEvent, StoppedEvent, BreakpointEvent, OutputEvent, Event,
+	Thread, StackFrame, Scope, Source, Handles, Breakpoint, ThreadEvent, Variable
+} from "vscode-debugadapter";
+import { DebugProtocol } from "vscode-debugprotocol";
+
+export interface DartLaunchRequestArguments extends DebugProtocol.LaunchRequestArguments {
+	cwd: string;
+	program: string;
+	args: Array<string>;
+}
+
+export class DartDebugSession extends DebugSession {
+	private sourceFile: string;
+	private cwd: string;
+	private childProcess: child_process.ChildProcess;
+	private processExited: boolean = false;
+
+	public constructor() {
+		super();
+
+		this.setDebuggerLinesStartAt1(false);
+		this.setDebuggerColumnsStartAt1(false);
+	}
+
+	protected initializeRequest(
+		response: DebugProtocol.InitializeResponse,
+		args: DebugProtocol.InitializeRequestArguments
+	): void {
+		response.body = {
+			supportsConfigurationDoneRequest: true,
+			supportsEvaluateForHovers: true,
+			exceptionBreakpointFilters: [
+				{ filter: "All", label: "All Exceptions", default: false },
+				{ filter: "Unhandled", label: "Uncaught Exceptions", default: true }
+			]
+		};
+		this.sendResponse(response);
+	}
+
+	protected launchRequest(response: DebugProtocol.LaunchResponse, args: DartLaunchRequestArguments): void {
+		this.cwd = args.cwd;
+		this.sourceFile = path.relative(args.cwd, args.program);
+		this.sendEvent(new OutputEvent(`dart ${this.sourceFile}\n`));
+
+		this.sendResponse(response);
+
+		let debug = !args.noDebug;
+		let appArgs = [];
+		// TODO:
+		// if (debug) {
+		// 	appArgs.push("--enable-vm-service:0");
+		// 	appArgs.push("--pause_isolates_on_start=true");
+		// }
+		appArgs.push(this.sourceFile);
+		if (args.args)
+			appArgs = appArgs.concat(args.args);
+
+		// TODO: dart sdk path - send a request to the client for it
```

Comment:
I don't think there's a way for the debug adapter to query the host vscode. The dart sdk location would have to be set in every launch config, which would be pretty annoying for users. We really want a way to pass a group of settings through automatically at launch (`dart.*`). The language servers have a feature like this -